processor/intervalprocessor/partitioned_processor.go

```diff
@@ -0,0 +1,284 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package intervalprocessor // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/intervalprocessor"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/intervalprocessor/internal/metrics"
+)
+
+type PartitionedProcessor struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	logger *zap.Logger
+
+	stateLock sync.Mutex
+
+	md                 []pmetric.Metrics
+	rmLookup           []map[identity.Resource]pmetric.ResourceMetrics
+	smLookup           []map[identity.Scope]pmetric.ScopeMetrics
+	mLookup            []map[identity.Metric]pmetric.Metric
+	numberLookup       []map[identity.Stream]pmetric.NumberDataPoint
+	histogramLookup    []map[identity.Stream]pmetric.HistogramDataPoint
+	expHistogramLookup []map[identity.Stream]pmetric.ExponentialHistogramDataPoint
+	summaryLookup      []map[identity.Stream]pmetric.SummaryDataPoint
```

Comment:
instead of having 8 slices that _must_ be used with the same index, can we have 1 struct slice?

```go
type Partition struct {
	md pmetric.Metrics
	rmLookup map[identity.Resource]pmetric.ResourceMetrics
	// ...
}

type Processor struct {
	// ...
	parts []Partition
}
```

would also eliminate `p.partitions`, as `len(p.parts)` has this info already